packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -74,123 +90,522 @@ const propTypes = {
+   */
+  const setDate = (inputValue, type) => {
+    if (type === DateUtil.inputType.DAY) {
+      dateDispatch({ day: inputValue });
+    } else if (type === DateUtil.inputType.MONTH) {
+      dateDispatch({ month: inputValue });
+    } else {
+      dateDispatch({ year: inputValue });
+    }
+
+    moveFocusOnChange(inputValue, type, dateFormatOrder);
+  };
 
-  handleOnChange(event) {
-    if (!DateUtil.validDateInput(event.target.value)) {
+  const handleDateChange = (event, inputValue, type) => {
+    /**
+     * Sets the day, month and year based on input values, formats them
+     * based on the date format variant, and passes the formatted date to onChange.
+     */
+    const day = type === DateUtil.inputType.DAY ? inputValue : date.day;
+    const month = type === DateUtil.inputType.MONTH ? inputValue : date.month;
+    const year = type === DateUtil.inputType.YEAR ? inputValue : date.year;
+
+    if (type === DateUtil.inputType.DAY) {
+      setDayInitialFocused(false);
+    } else if (type === DateUtil.inputType.MONTH) {
+      setMonthInitialFocused(false);
+    } else {
+      setYearInitialFocused(false);
+    }
+
+    let inputDate;
+    let formattedDate;
+    if (day.length === 2 && month.length === 2 && year.length === 4) {
+      inputDate = DateUtil.convertToISO8601(`${year}-${month}-${day}`, 'YYYY-MM-DD');
+      formattedDate = DateUtil.strictFormatISODate(inputDate, momentDateFormat);
+    }
+
+    if (onChange) {
+      if (DateUtil.isValidDate(formattedDate, momentDateFormat)) {
+        onChange(event, formattedDate);
+      } else {
+        let dateString;
+        if (dateFormatOrder === 'MDY') {
+          dateString = month + day + year;
+        } else if (dateFormatOrder === 'DMY') {
+          dateString = day + month + year;
+        } else {
+          dateString = year + month + day;
+        }
+        onChange(event, dateString);
+      }
+    }
+
+    setDate(inputValue, type);
+  };
+
+  const handleDayChange = (event) => {
+    let inputValue = event.target.value;
+    if (!DateUtil.validDateInput(inputValue)) {
       return;
     }
 
-    if (this.props.onChange) {
-      this.props.onChange(event);
+    // Ignore the entry if the value did not change or it is invalid.
+    // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have to
+    // check the length of inputValue to make sure that it is less then 2.
+    if (inputValue === date.day || inputValue.length > 2 || Number(inputValue) > 31 || inputValue === '00') {
+      return;
     }
-  }
 
-  handleOnKeyDown(event) {
-    if (this.props.onKeyDown) {
-      this.props.onKeyDown(event);
+    // If the change made was not a deletion of a digit, then prepend '0'
+    // when the input value is a single digit value between 4 and 9
+    if (inputValue.length >= date.day.length) {
+      const digitsToPrependZero = ['4', '5', '6', '7', '8', '9'];
```

Comment:
why not 0, 1, 2 or 3?

Reply:
So this logic has been separated into 3 parts:
1. The filtering logic based **on change** of the input element to determine whether or not the entered value is valid or not and auto prepend a zero accordingly. In this case entering a number between 4 to 9 will auto prepend a zero since we cannot have a day value beginning with these numbers.
2. In the case when a single "0" is entered, then we do not auto prepend a zero since 00 is not a valid value for day or month. The logic for that is [here](https://github.com/cerner/terra-framework/pull/1208/files/3ca87f6c56b6922eeb5da25cf3b48df7353dcf5f#diff-3636e37194def57a7d4ce937b69d965eR343)
3. In the case when the input is valid, in this case 1, 2 and 3 for day input, a zero is auto prepended only on navigating away from the respective input element. The logic for that is [here](https://github.com/cerner/terra-framework/pull/1208/files/3ca87f6c56b6922eeb5da25cf3b48df7353dcf5f#diff-3636e37194def57a7d4ce937b69d965eR478-R483) and handled **on blur** of the input element.

Reply:
It looks like entering just `0` for day or month and then lose focus will delete the zero. Is that correct? It seems a bit annoying that it delete it when the user may come back to finish the entry.